contributors/design-proposals/auth/external-secrets.md

```diff
@@ -0,0 +1,218 @@
+stored in etcd first. That might not be the ideal workflow because it
+does not support requirements like automatically renewing the
+certificate if the copy found in vault is getting close to expiration,
+or auditing the actual access of individual containerized applications
+to the secrets.
+
+It is possible to run per-Node daemon via DaemonSet with `hostPath`
+access to per-Node credentials (to authenticate to the external
+vault / secrets provider), exposing Unix socket on the filesystem of
+the Node. Using flexVolume mechanism, the directory with that Unix
+socket can then be made available to Pods and containers that need
+access to the secrets. The daemon in the DaemonSet can then use
+`SO_PEERCRED`/`SO_PEERSEC` to get the identity of the process accessing
+the service via the Unix socket, and authorize it based on the Pod
+identity and other attributes / annotations which can be retrieved
+via Kubernetes API.
+
+One drawback of this approach is the need for active client tool or
+library in each of the container, to invoke the daemon's API over
+the Unix socket to request the appropriate secret / certificate.
+
+For that reason we are looking for mechanism of exposing the secrets
+from external sources to Pods and their containers via filesystem
+interface. With that approach, the individual secrets would be seen as
+files in filesystem structure, while backed by active mechanisms
+of external secrets providers or vaults, and would only request the
+secret at the time it is actually accessed by the container.
+
+## Proposal
+
+We are proposing flexVolume plugin which would utilize FUSE. FUSE
+allows creating virtual filesystem where the content can be produced
+on-the-fly and behaviour of the individual entries can be actively
+controlled. The individual secrets would appear as files in the
+filesystem, while being generated (or even renewed, in case of
+certificates) via the backend logic, if needed.
+
+During `mount` operation, parameters like (formatted for clarity)
+
+```
+/var/lib/kubernetes/volumes/pods/215904af-a29b-11e7-a06b-5254005fe346/volumes/example.com~secrets-fuse/secrets-fuse
+{
+    "kubernetes.io/fsGroup":"1000020000",
+    "kubernetes.io/fsType":"",
+    "kubernetes.io/pod.name":"test-pod",
+    "kubernetes.io/pod.namespace":"default",
+    "kubernetes.io/pod.uid":"215904af-a29b-11e7-a06b-5254005fe346",
+    "kubernetes.io/pvOrVolumeName":"secrets-external",
+    "kubernetes.io/readwrite":"rw",
+    "kubernetes.io/serviceAccount.name":"default"
+}
+```
+
+are passed to the flexVolume plugin, so the flexVolume plugin can
+parse the parameters and get the `kubernetes.io/pod.uid` value
+and pass it to the FUSE filesystem via `mount` option. That provides
+the FUSE filesystem with the identity of the clients that will be
+accessing the secrets which can be used for access authorization.
+
+The FUSE filesystem runs as root so it can have access to all needed
```

Comment:
The container identity WG is currently proposing several options, including a private key created on the pods behalf that it may or may not have access to.  It would be good to frame this proposal in terms of the container identity WG stated goals and how this offers a near term mechanism to expose external secrets that is not dependent on the container identity.

Reply:
Do you have links to the latest proposals so that we could reference it here?

I'm not sure I understand the "external secrets that is not dependent on the container identity" -- is this "not dependent on container identity as opposed to pod identity", or "not dependent on container nor pod identity"? My understanding here is that the information / secrets that are made available are dependent on the pod identity.